### PR TITLE
Added SPR DMA & MFIFO support

### PR DESCRIPTION
**Pull request by refractionpcsx2**
Added SPR DMA & MFIFO support
Added GS FINISH interrupts
Added some missing DMA register read/writes
Added RSQRT Divide by Zero handling
Pass Emulator class to SPU2 for future interrupts

RoastMyPR.com :P

### CODE REVIEW

File: src/core/ee/dmac.cpp [hunk 1]
```diff
@@ -183,6 +187,33 @@ void DMAC::process_VIF0()
...
+    {
+        uint8_t id = (channels[VIF1].control >> 28) & 0x7;
+
+        channels[VIF1].tag_address = RBOR | (channels[VIF1].tag_address & RBSR);
+
+        //Don't mask the MADR on REFE/REF/REFS as they don't follow the tag, so likely outside the MFIFO
+        if (id != 0 && id != 3 && id != 4)
+        {
+            channels[VIF1].address = RBOR | (channels[VIF1].address & RBSR);
+        }
```
Comment (PSI-Rockin) [inline src/core/ee/dmac.cpp]: Don't use brackets for single-statement control flow blocks.

File: src/core/ee/dmac.cpp [hunk 2]
```diff
@@ -206,6 +237,33 @@ void DMAC::process_VIF1()
...
+    {
+        uint8_t id = (channels[GIF].control >> 28) & 0x7;
+
+        channels[GIF].tag_address = RBOR | (channels[GIF].tag_address & RBSR);
+
+        //Don't mask the MADR on REFE/REF/REFS as they don't follow the tag, so likely outside the MFIFO
+        if (id != 0 && id != 3 && id != 4)
+        {
+            channels[GIF].address = RBOR | (channels[GIF].address & RBSR);
+        }
```
Comment (PSI-Rockin) [inline src/core/ee/dmac.cpp]: Ditto. :P

File: src/core/ee/dmac.hpp [hunk 3]
```diff
@@ -11,6 +11,7 @@ struct DMA_Channel
     uint32_t quadword_count;
     uint32_t tag_address;
     uint32_t tag_save0, tag_save1;
+	uint32_t scratchpad_address;
```
Comment (PSI-Rockin) [inline src/core/ee/dmac.hpp]: Remove the excess indentation.

File: src/core/ee/vu.cpp [hunk 4]
```diff
@@ -465,10 +465,16 @@ void VectorUnit::rsqrt(uint8_t ftf, uint8_t fsf, uint8_t reg1, uint8_t reg2)
     if (!denom)
     {
         printf("[VU] RSQRT by zero!\n");
-        exit(1);
+        if ((gpr[reg1].u[fsf] & 0x80000000) != (gpr[reg2].u[ftf] & 0x80000000))
+            Q.u = 0xFF7FFFFF;
+        else
+            Q.u = 0x7F7FFFFF;
+    }
```
Comment (PSI-Rockin) [inline src/core/ee/vu.cpp]: I'm assuming this should set the status flags? See VectorUnit::div.

File: src/core/emulator.cpp [hunk 5]
```diff
@@ -378,6 +378,9 @@ uint32_t Emulator::read32(uint32_t address)
             return ipu.read_control();
         case 0x10002020:
             return ipu.read_BP();
+		case 0x10003C00:
+			printf("READ VIF1_STAT\n");
+			return 0;
```
Comment (PSI-Rockin) [inline src/core/emulator.cpp]: Excess indentation.

File: src/core/ee/vu.cpp (hunk 4 above)
Comment (refractionpcsx2) [reply src/core/ee/vu.cpp]: Yep, my bad, missed that, will update

File: src/core/ee/dmac.hpp (hunk 3 above)
Comment (refractionpcsx2) [reply src/core/ee/dmac.hpp]: For some reasaon it replaced that with a tab instead of spaces, I blame VS :P

**Pull request merged**